src/mime.cr

```diff
@@ -0,0 +1,287 @@
+require "crystal/system/mime"
+
+# This module implements a global MIME registry.
+#
+# ```
+# require "mime"
+#
+# MIME.from_extension(".html")         # => "text/html; charset=utf-8"
+# MIME.from_filename("path/file.html") # => "text/html; charset=utf-8"
+# ```
+#
+# The registry will be populated with some default values (see `DEFAULT_TYPES`)
+# as well as the operating system's MIME database.
+#
+# Default initialization can be skipped by calling `MIME.init(false)` before the first
+# query to the MIME database.
+#
+# ## OS-provided MIME database
+#
+# On a POSIX system, the following files are tried to be read in sequential order,
+# stopping at the first existing file. These values override those from `DEFAULT_TYPES`.
+#
+# ```plain
+# /etc/mime.types
+# /etc/httpd/mime.types                    # Mac OS X
+# /etc/httpd/conf/mime.types               # Apache
+# /etc/apache/mime.types                   # Apache 1
+# /etc/apache2/mime.types                  # Apache 2
+# /usr/local/etc/httpd/conf/mime.types
+# /usr/local/lib/netscape/mime.types
+# /usr/local/etc/httpd/conf/mime.types     # Apache 1.2
+# /usr/local/etc/mime.types                # FreeBSD
+# /usr/share/misc/mime.types               # OpenBSD
+# ```
+#
+# ## Registering custom MIME types
+#
+# Applications can register their own MIME types:
+#
+# ```
+# require "mime"
+# MIME.from_extension?(".cr")     # => nil
+# MIME.extensions("text/crystal") # => Set(String).new
+#
+# MIME.register(".cr", "text/crystal")
+# MIME.from_extension?(".cr")     # => "text/crystal"
+# MIME.extensions("text/crystal") # => Set(String){".cr"}
+# ```
+#
+# ## Loading a custom MIME database
+#
+# To load a custom MIME database, `load_mime_database` can be called with an
+# `IO` to read the database from.
+#
+# ```
+# # Load user-defined MIME types
+# File.open("~/.mime.types") do |io|
+#   MIME.load_mime_database(io)
+# end
+# ```
+#
+# Loaded values override previously defined mappings.
+#
+# The data format must follow the format of `mime.types`: Each line declares
+# a MIME type followed by a whitespace-separated list of extensions mapped to
+# this type. Everything following a `#` is considered a comment until the end of
+# line. Empy line are ignored.
+#
+# ```plain
+# text/html html htm
+#
+# # comment
+# ```
+module MIME
+  class Error < Exception
+  end
+
+  @@initialized = false
+  @@types = {} of String => String
+  @@types_lower = {} of String => String
+  @@extensions = {} of String => Set(String)
+
+  # A limited set of default MIME types.
+  DEFAULT_TYPES = {
+    ".css"  => "text/css; charset=utf-8",
+    ".gif"  => "image/gif",
+    ".htm"  => "text/html; charset=utf-8",
+    ".html" => "text/html; charset=utf-8",
+    ".jpg"  => "image/jpeg",
+    ".jpeg" => "image/jpeg",
+    ".js"   => "application/x-javascript",
```

Comment:
Adding `; charset=utf-8` IMO would be a good failsafe (same as in `.css`).

Reply:
Nope, see [here](https://mathiasbynens.be/demo/javascript-mime-type) for instance. Btw, official type for javascript is `application/javascript` (without `x-` prefix).